From aea5afcb03b19af37ed1c04cfdf294e9f4499f6b Mon Sep 17 00:00:00 2001
From: Matt Leon <mattleon@google.com>
Date: Wed, 16 Jul 2025 16:56:31 -0400
Subject: [PATCH 1/6] Restore _allowlist_function_transition

Reverts v8 commit b26554ec368e9553782012c96aa5e99b163eaff2, which removed use of
_allowlist_function_transition from v8 bazel/defs.bzl, since it is still required
by the version of Bazel we currently use (6.5.0).

Signed-off-by: Matt Leon <mattleon@google.com>
---
 bazel/defs.bzl                       |  3 +++
 bazel/v8-non-pointer-compression.bzl | 11 +++++++++++
 2 files changed, 14 insertions(+)

diff --git a/bazel/defs.bzl b/bazel/defs.bzl
index a7e1e2af786..ba8315bb938 100644
--- a/bazel/defs.bzl
+++ b/bazel/defs.bzl
@@ -491,6 +491,9 @@ _v8_mksnapshot = rule(
             cfg = "exec",
         ),
         "target_os": attr.string(mandatory = True),
+        "_allowlist_function_transition": attr.label(
+            default = "@bazel_tools//tools/allowlists/function_transition_allowlist",
+        ),
         "prefix": attr.string(mandatory = True),
         "suffix": attr.string(mandatory = True),
     },
diff --git a/bazel/v8-non-pointer-compression.bzl b/bazel/v8-non-pointer-compression.bzl
index 8c929454840..57336154cf7 100644
--- a/bazel/v8-non-pointer-compression.bzl
+++ b/bazel/v8-non-pointer-compression.bzl
@@ -47,6 +47,17 @@ v8_binary_non_pointer_compression = rule(
         # Note specificaly how it's configured with v8_target_cpu_transition, which
         # ensures that setting propagates down the graph.
         "binary": attr.label(cfg = v8_disable_pointer_compression),
+        # This is a stock Bazel requirement for any rule that uses Starlark
+        # transitions. It's okay to copy the below verbatim for all such rules.
+        #
+        # The purpose of this requirement is to give the ability to restrict
+        # which packages can invoke these rules, since Starlark transitions
+        # make much larger graphs possible that can have memory and performance
+        # consequences for your build. The whitelist defaults to "everything".
+        # But you can redefine it more strictly if you feel that's prudent.
+        "_allowlist_function_transition": attr.label(
+            default = "@bazel_tools//tools/allowlists/function_transition_allowlist",
+        ),
     },
     # Making this executable means it works with "$ bazel run".
     executable = True,
-- 
2.52.0.457.g6b5491de43-goog


From 920ef0496317a512083a09caa905fbf87a01c14b Mon Sep 17 00:00:00 2001
From: Matt Leon <mattleon@google.com>
Date: Wed, 16 Jul 2025 16:56:52 -0400
Subject: [PATCH 2/6] Don't expose Wasm C API (only Wasm C++ API).

Signed-off-by: Matt Leon <mattleon@google.com>
---
 src/wasm/c-api.cc | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/src/wasm/c-api.cc b/src/wasm/c-api.cc
index e7cf656ade4..660ed06021d 100644
--- a/src/wasm/c-api.cc
+++ b/src/wasm/c-api.cc
@@ -2481,6 +2481,8 @@ WASM_EXPORT auto Instance::exports() const -> ownvec<Extern> {
 
 }  // namespace wasm
 
+#if 0
+
 // BEGIN FILE wasm-c.cc
 
 extern "C" {
@@ -3527,3 +3529,5 @@ wasm_instance_t* wasm_frame_instance(const wasm_frame_t* frame) {
 #undef WASM_DEFINE_SHARABLE_REF
 
 }  // extern "C"
+
+#endif
-- 
2.52.0.457.g6b5491de43-goog


From f6379d9391b775f0459c3740302e435421a942c1 Mon Sep 17 00:00:00 2001
From: Matt Leon <mattleon@google.com>
Date: Wed, 16 Jul 2025 16:58:02 -0400
Subject: [PATCH 3/6] Stub out fast_float for bazel-supplied version

Signed-off-by: Matt Leon <mattleon@google.com>
---
 BUILD.bazel | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/BUILD.bazel b/BUILD.bazel
index 96c8be8901f..05630ab7035 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -4557,7 +4557,7 @@ v8_library(
     ],
     deps = [
         ":lib_dragonbox",
-        "//third_party/fast_float/src:fast_float",
+        "@fast_float//:fast_float",
         ":lib_fp16",
         ":simdutf",
         ":v8_libbase",
-- 
2.52.0.457.g6b5491de43-goog


From 98a46e051afa5186a07400c53923367b9bea0ebe Mon Sep 17 00:00:00 2001
From: Matt Leon <mattleon@google.com>
Date: Wed, 16 Jul 2025 20:04:05 -0400
Subject: [PATCH 4/6] Stub out vendored dependencies for bazel-sourced versions

Signed-off-by: Matt Leon <mattleon@google.com>
---
 BUILD.bazel | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/BUILD.bazel b/BUILD.bazel
index 05630ab7035..7f14a987148 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -4556,10 +4556,10 @@ v8_library(
         ":noicu/generated_torque_definitions",
     ],
     deps = [
-        ":lib_dragonbox",
+        "@dragonbox//:dragonbox",
         "@fast_float//:fast_float",
-        ":lib_fp16",
-        ":simdutf",
+        "@fp16//:FP16",
+        "@simdutf//:simdutf",
         ":v8_libbase",
         "@abseil-cpp//absl/container:btree",
         "@abseil-cpp//absl/container:flat_hash_map",
-- 
2.52.0.457.g6b5491de43-goog


From e3cebe8ca694935302a8c3806ab6398c593f31cd Mon Sep 17 00:00:00 2001
From: Matt Leon <mattleon@google.com>
Date: Wed, 16 Jul 2025 20:29:10 -0400
Subject: [PATCH 5/6] Add build flags to make V8 compile with GCC

Signed-off-by: Matt Leon <mattleon@google.com>
---
 bazel/defs.bzl | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/bazel/defs.bzl b/bazel/defs.bzl
index ba8315bb938..277a07ef63d 100644
--- a/bazel/defs.bzl
+++ b/bazel/defs.bzl
@@ -123,6 +123,9 @@ def _default_args():
                 "-Wno-deprecated-copy",
                 "-Wno-non-virtual-dtor",
                 "-Wno-unnecessary-virtual-specifier",
+                "-Wno-invalid-offsetof",
+                "-Wno-dangling-pointer",
+                "-Wno-dangling-reference",
                 "-isystem .",
             ],
             "//conditions:default": [],
-- 
2.52.0.457.g6b5491de43-goog


From 5f9cd8cf1b1758678d41f96cd22477e052c787f8 Mon Sep 17 00:00:00 2001
From: Matt Leon <mattleon@google.com>
Date: Fri, 18 Jul 2025 17:28:42 -0400
Subject: [PATCH 6/6] Hack out atomic simd support in V8.

Atomic simdutf requires __cpp_lib_atomic_ref >= 201806, which is only
supported in clang libc++ 19+. The version of LLVM used in Envoy as of
2025-07-18 is libc++ 18, so this is not supported.

The simdutf documentation indicates this atomic form is not tested and
is not recommended for use:
https://github.com/simdutf/simdutf/blob/5d1b6248f29a8ed0eb90f79be268be41730e39f8/include/simdutf/implementation.h#L3066-L3068

In addition, this is in the implementation of a JS array buffer. Since
proxy-wasm-cpp-host does not make use of JS array buffers or shared
memory between web workers, we're stubbing it out.

Mostly reverts
https://github.com/v8/v8/commit/6d6c1e680c7b8ea5f62a76e9c3d88d3fb0a88df0.

Signed-off-by: Matt Leon <mattleon@google.com>
---
 bazel/defs.bzl                       |  2 +-
 src/builtins/builtins-typed-array.cc |  8 ++++++++
 src/objects/simd.cc                  | 10 ++++++++++
 3 files changed, 19 insertions(+), 1 deletion(-)

diff --git a/bazel/defs.bzl b/bazel/defs.bzl
index 277a07ef63d..0617ecf1607 100644
--- a/bazel/defs.bzl
+++ b/bazel/defs.bzl
@@ -186,7 +186,7 @@ def _default_args():
                 "Advapi32.lib",
             ],
             "@v8//bazel/config:is_macos": ["-pthread"],
-            "//conditions:default": ["-Wl,--no-as-needed -ldl -latomic -pthread"],
+            "//conditions:default": ["-Wl,--no-as-needed -ldl -pthread"],
         }) + select({
             ":should_add_rdynamic": ["-rdynamic"],
             "//conditions:default": [],
diff --git a/src/builtins/builtins-typed-array.cc b/src/builtins/builtins-typed-array.cc
index af505a1f729..8c41631f291 100644
--- a/src/builtins/builtins-typed-array.cc
+++ b/src/builtins/builtins-typed-array.cc
@@ -506,17 +506,21 @@ simdutf::result ArrayBufferSetFromBase64(
     DirectHandle<JSTypedArray> typed_array, size_t& output_length) {
   output_length = array_length;
   simdutf::result simd_result;
+#ifdef WANT_ATOMIC_REF
   if (typed_array->buffer()->is_shared()) {
     simd_result = simdutf::atomic_base64_to_binary_safe(
         reinterpret_cast<const T>(input_vector), input_length,
         reinterpret_cast<char*>(typed_array->DataPtr()), output_length,
         alphabet, last_chunk_handling, /*decode_up_to_bad_char*/ true);
   } else {
+#endif
     simd_result = simdutf::base64_to_binary_safe(
         reinterpret_cast<const T>(input_vector), input_length,
         reinterpret_cast<char*>(typed_array->DataPtr()), output_length,
         alphabet, last_chunk_handling, /*decode_up_to_bad_char*/ true);
+#ifdef WANT_ATOMIC_REF
   }
+#endif
 
   return simd_result;
 }
@@ -840,15 +844,19 @@ BUILTIN(Uint8ArrayPrototypeToBase64) {
     // 11. Return CodePointsToString(outAscii).
 
     size_t simd_result_size;
+#ifdef WANT_ATOMIC_REF
     if (uint8array->buffer()->is_shared()) {
       simd_result_size = simdutf::atomic_binary_to_base64(
           reinterpret_cast<const char*>(uint8array->DataPtr()), length,
           reinterpret_cast<char*>(output->GetChars(no_gc)), alphabet);
     } else {
+#endif
       simd_result_size = simdutf::binary_to_base64(
           reinterpret_cast<const char*>(uint8array->DataPtr()), length,
           reinterpret_cast<char*>(output->GetChars(no_gc)), alphabet);
+#ifdef WANT_ATOMIC_REF
     }
+#endif
     DCHECK_EQ(simd_result_size, output_length);
     USE(simd_result_size);
   }
diff --git a/src/objects/simd.cc b/src/objects/simd.cc
index ef2b8118eec..68115031f1a 100644
--- a/src/objects/simd.cc
+++ b/src/objects/simd.cc
@@ -477,6 +477,7 @@ void Uint8ArrayToHexSlow(const char* bytes, size_t length,
   }
 }
 
+#ifdef WANT_ATOMIC_REF
 void AtomicUint8ArrayToHexSlow(const char* bytes, size_t length,
                                DirectHandle<SeqOneByteString> string_output) {
   int index = 0;
@@ -492,6 +493,7 @@ void AtomicUint8ArrayToHexSlow(const char* bytes, size_t length,
     index += 2;
   }
 }
+#endif
 
 inline uint16_t ByteToHex(uint8_t byte) {
   const uint16_t correction = (('a' - '0' - 10) << 8) + ('a' - '0' - 10);
@@ -647,11 +649,15 @@ Tagged<Object> Uint8ArrayToHex(const char* bytes, size_t length, bool is_shared,
   }
 #endif
 
+#ifdef WANT_ATOMIC_REF
   if (is_shared) {
     AtomicUint8ArrayToHexSlow(bytes, length, string_output);
   } else {
+#endif
     Uint8ArrayToHexSlow(bytes, length, string_output);
+#ifdef WANT_ATOMIC_REF
   }
+#endif
   return *string_output;
 }
 
@@ -1084,12 +1090,16 @@ bool ArrayBufferFromHex(const base::Vector<T>& input_vector, bool is_shared,
   for (uint32_t i = 0; i < output_length * 2; i += 2) {
     result = HandleRemainingHexValues(input_vector, i);
     if (result.has_value()) {
+#ifdef WANT_ATOMIC_REF
       if (is_shared) {
         std::atomic_ref<uint8_t>(buffer[index++])
             .store(result.value(), std::memory_order_relaxed);
       } else {
+#endif
         buffer[index++] = result.value();
+#ifdef WANT_ATOMIC_REF
       }
+#endif
     } else {
       return false;
     }
-- 
2.52.0.457.g6b5491de43-goog

